src/pages/Group/GroupFeed.tsx

```diff
@@ -0,0 +1,184 @@
+    queryFn: () =>
+      api
+        .get<Happiness[]>(`/group/${groupData.id}/happiness/unread`)
+        .then((res) => res.data),
+    cacheTime: 0, // disable cache since this causes stale reads to be shown on initial load
+  });
+
+  // infinite query for fetching regular entries
+  const feedQuery = useInfiniteQuery<HappinessPagination>({
+    queryKey: [QueryKeys.FETCH_GROUP_HAPPINESS, groupData.id],
+    queryFn: ({ pageParam = 1 }) =>
+      api
+        .get<Happiness[]>(`/group/${groupData.id}/happiness/count`, {
+          page: pageParam,
+        })
+        .then((res): HappinessPagination => {
+          return {
+            data: res.data,
+            page: pageParam,
+          };
+        }),
+    getNextPageParam: (lastPage: HappinessPagination) => {
+      // no more pages left if the last page is empty
+      if (lastPage.data.length === 0) return false;
+      return lastPage.page + 1; // increment page number to fetch
+    },
+  });
+
+  // flatten React query pagination object into single list
+  const allFeedEntries = useMemo(
+    () =>
+      feedQuery.data?.pages.reduce(
+        (acc: Happiness[], page) => [...acc, ...page.data],
+        [],
+      ),
+    [feedQuery.data],
+  );
+
+  const entryToJsx = (
+    entry: Happiness,
+    prevEntry: Happiness | undefined,
+    unreadReq: boolean,
+  ) => {
+    const feedCard = (
+      <FeedCard
+        key={entry.id * (unreadReq ? -1 : 1)}
+        data={entry}
+        isNew={
+          unreadReq ||
+          unreadQuery.data!.findIndex((elt) => elt.id == entry.id) !== -1
+        }
+        onClick={() => setSelectedEntry(entry)}
+        trackRead={unreadReq}
+      />
+    );
+
+    // add date boundary if the previous entry has a different timestamp
+    if (prevEntry === undefined || entry.timestamp !== prevEntry.timestamp) {
+      return (
+        <div key={entry.timestamp + " " + unreadReq}>
+          <h5 className="mb-4 text-gray-400">
+            {dateOrTodayYesterday(
+              entry.timestamp,
+              dateFromStr(entry.timestamp).toLocaleDateString("en-us", {
+                year: "2-digit",
+                month: "2-digit",
+                day: "2-digit",
+              }),
+            )}
+          </h5>
+          {feedCard}
+        </div>
+      );
+    }
+    return feedCard;
+  };
+
+  // load more entries when bottom of infinite scroll reached
+  useEffect(() => {
+    if (bottomInView && feedQuery.hasNextPage) feedQuery.fetchNextPage();
+  }, [bottomInView]);
+
+  return (
+    <div className="mx-8">
+      {/* Unread feed */}
+      {unreadQuery.isLoading ? (
+        <Spinner />
+      ) : (
+        <>
+          {unreadQuery.isError ? (
+            <h5 className="text-gray-400">Error: Could not load entries.</h5>
+          ) : (
+            <>
+              {unreadQuery.data!.length > 0 && <h5 className="mb-4">Unread</h5>}
+              {unreadQuery.data!.map((entry, i) =>
```

Comment:
I think you should use `data?` instead of `data!` so this way if the data is undefined then instead of the app crashing it will just not show unread entries. Make sure to change all of these non-null assertions to the `?` variations in this file.

Reply:
i don't think that happens? data is guaranteed to be defined at that portion of the code since we check to make sure the query is not in the loading or error state. also if there are no unread entries data will just be []